cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+                    if not self.dependency_dict[bear]:
+                        return_bears.append(bear)
+        for bear in return_bears:
+            del self.dependency_dict[bear]
+        return return_bears
+```
+
+# A Prototype
+
+A rough look at how the new local bear class implementation will look:
+
+```python
+class FileWiseBear:
+    def __init__(self, section, result_queue, file_set):
+        """
+        Constructs a new bear.
+
+        :param section:      The section object where bear settings are
+                             contained.
+        :param result_queue: The queue object for messages. Can be ``None``.
+        :param file_set:     A set of FileProxy objects.
+        """
+        self.section = section
+        self.result_queue = result_queue
+        self.file_set = file_set
+
+        # May raise RuntimeError so bear doesn't get executed on invalid params
+        self.kwargs = get_kwargs_for_function(self.run, section)
+
+    def execute_task(self, task, **kwargs):
+        """
+        This method is responsible for getting results from the
+        analysis routines and populating the result queue.
+        """
+        return list(self.analyze(task, **self.kwargs))
+
+    def analyze(self,
+                file_proxy,
+                *args,
+                dependency_results=None,
+                **kwargs):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        raise NotImplementedError("This function has to be implemented for a "
+                                  "runnable bear.")
+
+    def generate_tasks(self):
+        """
+        This method is responsible for providing the files.
+        """
+        return self.file_set
+```
+
+A test bear:
+
+```python
+class TestBear(FileWiseBear):
+    def analyze(self, file_proxy):
+        """
+        Handles the given file.
+
+        :param file_proxy: Object containing filename and contents
+        :return:           A list of Result
+        """
+        for i, line in enumerate(file):
+            # This will contain the analysis routine
+            yield "Line has a problem" , file_proxy.name, i
+```
+
+A rough idea of how the new bear will be called:
+
+```python
+import asyncio
+import concurrent.futures
+import logging
+import sys
+from coalib.collecting.Collectors import collect_files
+from Fileproxy import FileProxy
+from TestBear import TestBear
+from AnotherBear import AnotherBear
+from Dependencies import Dependencies
+from DependentBear import DependentBear
+from AnotherDependentBear import AnotherDependentBear
+import functools
+import multiprocessing
+
+
+def get_cpu_count():
+    try:
+        return multiprocessing.cpu_count()
+    # cpu_count is not implemented for some CPU architectures/OSes
+    except NotImplementedError:  # pragma: no cover
+        return 2
```

Comment:
I think we can now return just `1` because we don't need another logger thread (I think that was the reason providing `2`... I believe...). Correct me if I'm wrong :P